tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/OsgiBundleProject.java

```diff
@@ -241,20 +240,19 @@ private void addPDESourceRoots(MavenProject project) {
         EclipsePluginProject eclipsePluginProject = getEclipsePluginProject(DefaultReactorProject.adapt(project));
         for (BuildOutputJar outputJar : eclipsePluginProject.getOutputJars()) {
             for (File sourceFolder : outputJar.getSourceFolders()) {
-                removeDuplicateTestCompileRoot(sourceFolder, project.getTestCompileSourceRoots());
+                removeDuplicateTestCompileRoot(sourceFolder, project);
                 project.addCompileSourceRoot(sourceFolder.getAbsolutePath());
             }
         }
     }
 
-    private void removeDuplicateTestCompileRoot(File sourceFolder, List<String> testCompileSourceRoots) {
-        for (Iterator<String> iterator = testCompileSourceRoots.iterator(); iterator.hasNext();) {
-            String testCompileRoot = iterator.next();
+    private void removeDuplicateTestCompileRoot(File sourceFolder, MavenProject project) {
+        for (String testCompileRoot : project.getTestCompileSourceRoots()) {
```

Comment:
I think we must first copy the list, the implicit iterator of the for-each loop does not play well together with calling remove() on the same list instance.

```diff
-for (String testCompileRoot : project.getTestCompileSourceRoots()) {
+for (String testCompileRoot : new ArrayList<>(project.getTestCompileSourceRoots())) {
```

Reply:
@sratz while this is true in general and I had the same feeling before when adapting the code, we immediately return when finding a match here so it is okay in this case.